### PR TITLE
Using bootstrap options sent via azure while bootstrapping

### PR DESCRIPTION
**Pull request by NimishaS**
Passing bootstrap options send via azure to chef extension for bootstrapping on windows and linux

### CODE REVIEW

File: lib/chef/azure/commands/enable.rb [hunk 1]
```diff
@@ -117,27 +120,44 @@ def configure_chef_only_once
...
-      end
-      # write the first_boot.json
-      File.open("#{bootstrap_directory}/first-boot.json", "w") do |f|
-        f.write(<<-RUNLIST
-{
-"run_list": [#{ @run_list.empty? ? "" : escape_runlist(@run_list)}]
-}
-RUNLIST
-)
+      if windows?
```
Comment (adamedx) [inline lib/chef/azure/commands/enable.rb]: Why is this conditional on Windows? Shouldn't we do the same thing on all platforms?

File: lib/chef/azure/commands/enable.rb [hunk 2]
```diff
@@ -117,27 +120,44 @@ def configure_chef_only_once
...
 
       # run chef-client for first time with no runlist to register the node
       puts "Running chef client for first time with no runlist..."
 
       begin
-        params = " -c #{bootstrap_directory}/client.rb -E _default -L #{@azure_plugin_log_location}/chef-client.log --once "
-        result = shell_out("chef-client #{params}")
-        result.error!
+        if windows?
+          params = " -c #{bootstrap_directory}/client.rb -E _default -L #{@azure_plugin_log_location}/chef-client.log --once "
```
Comment (adamedx) [inline lib/chef/azure/commands/enable.rb]: So for non-Windows I see the new options, but why don't we also do this for Windows?

Comment (NimishaS) [inline lib/chef/azure/commands/enable.rb]: For windows we need windows bootstrap context and template from knife-windows. So we are implementing this for linux first.

**Comment by NimishaS** [comment]
This is in progress

File: lib/chef/azure/commands/enable.rb [hunk 3]
```diff
@@ -112,31 +113,38 @@ def configure_chef_only_once
...
+        config[:chef_node_name] = value_from_json_file(handler_settings_file,'runtimeSettings','0','handlerSettings', 'publicSettings', 'bootstrap_options','chef_node_name')
+        config[:encrypted_data_bag_secret ] = value_from_json_file(handler_settings_file,'runtimeSettings','0','handlerSettings', 'publicSettings', 'bootstrap_options','encrypted_data_bag_secret')
+        config[:chef_extension_root] = @chef_extension_root
+        Chef::Config[:validation_key_content] = @validation_key
+        Chef::Config[:chef_server_url] = value_from_json_file(handler_settings_file,'runtimeSettings','0','handlerSettings', 'publicSettings', 'bootstrap_options','chef_server_url')
+        Chef::Config[:validation_client_name] = value_from_json_file(handler_settings_file,'runtimeSettings','0','handlerSettings', 'publicSettings', 'bootstrap_options','validation_client_name')
+
+        if windows?
+          context = Chef::Knife::Core::WindowsBootstrapContext.new(config, {}, Chef::Config)
+          template_file = Gem.find_files(File.join("chef","azure","bootstrap","windows-chef-client-msi.erb")).first
```
Comment (siddheshwar-more) [inline lib/chef/azure/commands/enable.rb]: Is it a right way to find file path? I think Gem.find_files used to find files in other gems?

File: lib/chef/azure/commands/enable.rb [hunk 4]
```diff
@@ -112,31 +113,38 @@ def configure_chef_only_once
...
+          bootstrap_bat_file ||= "#{ENV['TMP']}/bootstrap.bat"
```
Comment (siddheshwar-more) [inline lib/chef/azure/commands/enable.rb]: Can you append time-stamp to temp bootstrap_bat_file ?

File: lib/chef/azure/core/windows_bootstrap_context.rb [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+        def encrypted_data_bag_secret
+          escape_and_echo(@config[:encrypted_data_bag_secret])
+        end
+
+        def trusted_certs_script
+          @trusted_certs_script ||= trusted_certs_content
+        end
+
+        def config_content
```
Comment (siddheshwar-more) [inline lib/chef/azure/core/windows_bootstrap_context.rb]: can we move duplicate code to base `lib/chef/azure/core/bootstrap_context.rb` template ?

File: lib/chef/azure/core/bootstrap_context.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        def start_chef
+          # If the user doesn't have a client path configure, let bash use the PATH for what it was designed for
+          client_path = @chef_config[:chef_client_path] || 'chef-client'
+          s = "#{client_path} "
+          s << ' -l debug' if @config[:verbosity] and @config[:verbosity] >= 2
+          s << " -E #{bootstrap_environment}"
+          s
+        end
+
+        def config_content
```
Comment (siddheshwar-more) [inline lib/chef/azure/core/bootstrap_context.rb]: Does this take cares of user's client_rb file? Ideally we should append client_rb content to user provided client_rb file!

File: lib/chef/azure/commands/enable.rb [hunk 7]
```diff
@@ -112,31 +113,38 @@ def configure_chef_only_once
...
       load_settings
 
-      # Write validation key
-      File.open("#{bootstrap_directory}/validation.pem", "w") do |f|
-        f.write(@validation_key)
-      end
-
-      # Write client.rb
-      File.open("#{bootstrap_directory}/client.rb", "w") do |f|
-        f.write(override_clientrb_file(@client_rb))
```
Comment (siddheshwar-more) [inline lib/chef/azure/commands/enable.rb]: Can you remove `def override_clientrb_file(client_rb)` method If its not used anywhere ?

File: lib/chef/azure/commands/enable.rb (hunk 4 above)
Comment (NimishaS) [inline lib/chef/azure/commands/enable.rb]: We are deleting this file immediately. So it won't be referred later.

**Comment by NimishaS** [comment]
Fixed all review comments by @siddheshwar-more

File: lib/chef/azure/commands/enable.rb [hunk 8]
```diff
@@ -112,31 +113,41 @@ def configure_chef_only_once
...
-      end
-
       # run chef-client for first time with no runlist to register the node
       puts "Running chef client for first time with no runlist..."
 
       begin
-        params = " -c #{bootstrap_directory}/client.rb -E _default -L #{@azure_plugin_log_location}/chef-client.log --once "
-        result = shell_out("chef-client #{params}")
+        require 'chef/azure/core/bootstrap_context'
+        config = {}
```
Comment (adamedx) [inline lib/chef/azure/commands/enable.rb]: This appears to be where we've added new json options to the extension. We should update the README to document the json format.

**Pull request merged**